common/lib/xmodule/xmodule/tabs.py

```diff
@@ -174,62 +169,104 @@ def from_json(tab_dict):
+        from openedx.core.djangoapps.plugins.api import CourseViewType
+        if issubclass(tab_type, CourseViewType):
+            return CourseViewTab(tab_type, tab_dict=tab_dict)
+        else:
+            return tab_type(tab_dict=tab_dict)
 
-        tab_class = sub_class_types[tab_dict['type']]
-        tab_class.validate(tab_dict)
-        return tab_class(tab_dict=tab_dict)
+
+class CourseTabManager(object):
+    """
+    A manager that handles the set of available course tabs.
+    """
+
+    @classmethod
+    def get_tab_types(cls):
+        """
+        Returns the list of available tab types.
+        """
+        if not hasattr(cls, "_tab_types"):
+            tab_types = {
+                'courseware': CoursewareTab,
+                'course_info': CourseInfoTab,
+                'wiki': WikiTab,
+                'discussion': DiscussionTab,
+                'external_discussion': ExternalDiscussionTab,
+                'external_link': ExternalLinkTab,
+                'textbooks': TextbookTabs,
+                'pdf_textbooks': PDFTextbookTabs,
+                'html_textbooks': HtmlTextbookTabs,
+                'progress': ProgressTab,
+                'static_tab': StaticTab,
+                'peer_grading': PeerGradingTab,
+                'staff_grading': StaffGradingTab,
+                'open_ended': OpenEndedGradingTab,
+                'notes': NotesTab,
+                'syllabus': SyllabusTab,
+            }
+
+            # Add any registered course views
+            # TODO: don't import openedx capabilities from common
+            from openedx.core.djangoapps.plugins.api import CourseViewTypeManager
+            for course_view_type in CourseViewTypeManager.get_available_plugins().values():
+                tab_types[course_view_type.name] = course_view_type
+
+            cls._tab_types = tab_types
+        return cls._tab_types
 
 
 class AuthenticatedCourseTab(CourseTab):
     """
     Abstract class for tabs that can be accessed by only authenticated users.
     """
-    def can_display(self, course, settings, is_user_authenticated, is_user_staff, is_user_enrolled):
-        return is_user_authenticated
+    def is_enabled(self, course, settings, user=None):
+        return not user or user.is_authenticated()
 
 
 class StaffTab(AuthenticatedCourseTab):
     """
     Abstract class for tabs that can be accessed by only users with staff access.
     """
-    def can_display(self, course, settings, is_user_authenticated, is_user_staff, is_user_enrolled):  # pylint: disable=unused-argument
-        return is_user_staff
+    def is_enabled(self, course, settings, user=None):  # pylint: disable=unused-argument
+        return not user or is_user_staff(course, user)
 
 
-class EnrolledOrStaffTab(CourseTab):
+def is_user_staff(course, user):
+    """
+    Returns true if the user is staff in the specified course, or globally.
+    """
+    from courseware.access import has_access  # pylint: disable=import-error
```

Comment:
This should be raising large warning flags. If you have to import inside a method, then something is probably wrong.